metrics.go

```diff
@@ -59,25 +81,34 @@ func newPrometheusScope(c prometheus.Configuration, prefix string, log *zap.Logg
 	return scope, closer, nil
 }
 
-func (p *Plugin) MetricsCollector() []prom.Collector {
-	// p - implements Exporter interface (workers)
-	// other - request duration and count
-	return []prom.Collector{p.statsExporter}
-}
+// ref: https://github.dev/temporalio/temporal/common/metrics/config.go:391
+func newStatsdScope(statsdConfig *Statsd) (tally.Scope, io.Closer, error) {
+	st, err := statsd.NewClientWithConfig(&statsd.ClientConfig{
+		Address:       statsdConfig.HostPort,
+		Prefix:        statsdConfig.Prefix,
+		UseBuffered:   false,
+		FlushInterval: statsdConfig.FlushInterval,
+		FlushBytes:    statsdConfig.FlushBytes,
+	})
 
-const (
-	namespace = "rr_temporal"
-)
+	if err != nil {
+		return nil, nil, err
+	}
 
-func newStatsExporter(stats informer.Informer) *metrics.StatsExporter {
-	return &metrics.StatsExporter{
-		TotalMemoryDesc:  prom.NewDesc(prom.BuildFQName(namespace, "", "workers_memory_bytes"), "Memory usage by workers", nil, nil),
-		StateDesc:        prom.NewDesc(prom.BuildFQName(namespace, "", "worker_state"), "Worker current state", []string{"state", "pid"}, nil),
-		WorkerMemoryDesc: prom.NewDesc(prom.BuildFQName(namespace, "", "worker_memory_bytes"), "Worker current memory usage", []string{"pid"}, nil),
-		TotalWorkersDesc: prom.NewDesc(prom.BuildFQName(namespace, "", "total_workers"), "Total number of workers used by the plugin", nil, nil),
-		WorkersReady:     prom.NewDesc(prom.BuildFQName(namespace, "", "workers_ready"), "Workers currently in ready state", nil, nil),
-		WorkersWorking:   prom.NewDesc(prom.BuildFQName(namespace, "", "workers_working"), "Workers currently in working state", nil, nil),
-		WorkersInvalid:   prom.NewDesc(prom.BuildFQName(namespace, "", "workers_invalid"), "Workers currently in invalid,killing,destroyed,errored,inactive states", nil, nil),
-		Workers:          stats,
+	// NOTE: according to (https://github.com/uber-go/tally) Tally's statsd implementation doesn't support tagging.
+	// Therefore, we implement Tally interface to have a statsd reporter that can support tagging
+	opts := statsdreporter.Options{
+		TagSeparator: statsdConfig.Reporter.TagSeparator,
```

Comment:
@rustatian First of all bit thanks for putting effort here, really nice job!

1. From the pragmatic point of view, Temporal implementation of statsd without tagging is a bit useless (I mean like tags used for prometheus). The thing is as a user I should have possibility to choice - I am using tags or combination tags in names. 

E.g. with Datadog there is a big problem to create good metrics based on names combined with tags. 

There is workaround for Temporal, but tbh I don't want to fork temporal-rr to keep such functionallity alive because is double work for me.

https://docs.temporal.io/references/server-options/#withcustommetricsreporter
https://github.com/temporalio/temporal/pull/998#issuecomment-857884983

2. Would be nice to have possibility own tags. E.g. I would like to add tag "app_env". In most cases people using same observing tools for stg and production and they recognise everything based on tags.

Reply:
Hey @cv65kr 👋🏻 
I looked at the temporal implementation of statsd, and actually, it supports tags: https://github.com/temporalio/temporal/blob/master/common/metrics/tally/statsd/reporter.go
As stated in the second part of the comment:
`// Therefore, we implement Tally interface to have a statsd reporter that can support tagging`

Reply:
Hey @rustatian 
Unfortunately concrete implementation of statsd used in Temporal doesn't use tags functionality. You can use interface and implement your own driver - https://github.com/temporalio/temporal/pull/998#issuecomment-857884983

Reply:
The tags are injected into a name in the temporal std implementation. I'll go with the current implementation at the moment to support a statsd. You may try it and send a different feature request if that doesn't help you (but to the RR repo: https://github.com/roadrunner-server/roadrunner).

That sample is from 2021, and IIRC, Tally and datadog-statsd had updates in their API. So, I need time to adapt it appropriately.

Reply:
yes, unfortunately with tags injected in name is quite hard to create proper dashboards in Datadog.

I can send you my solution for DD and newest version of temporal if you want.

Reply:
Sure, you may attach your files to a new ticket with Tags FR 👍🏻

Reply:
Generally it's only about starting Temporal, the implementation is still the same:
```
// Reporter is from link which I sent earlier
reporter := NewReporter(&DogstatsdReporterConfig{HostPort: cfg.Global.Metrics.Statsd.HostPort}, logger)
				scopeOpts := tally.ScopeOptions{
					Tags:     cfg.Global.Metrics.Tags,
					Reporter: reporter,
					Prefix:   "temporal_dd",
				}
				scope, _ := tally.NewRootScope(scopeOpts, time.Second)
				tallyMetricsHandler := metrics.NewTallyMetricsHandler(cfg.Global.Metrics.ClientConfig, scope)

				s := temporal.NewServer(
					temporal.WithCustomMetricsHandler(tallyMetricsHandler),
				)
```

Reply:
As far as I see, `datadog` uses a data-dog specific format for the metrics: https://docs.datadoghq.com/developers/dogstatsd/datagram_shell?tab=metrics
`<METRIC_NAME>:<VALUE>|<TYPE>|@<SAMPLE_RATE>|#<TAG_KEY_1>:<TAG_VALUE_1>,<TAG_2>`

So, if we go with the proposed implementation, we bind to the datadog only. So, that definitely should be a separate FR. I need to think about how to do that in a generic way.

Reply:
The client used in our implementation has a formatter for the different implementations: https://github.com/cactus/go-statsd-client/blob/master/statsd/client_config.go#L47

So, I'll prefer to implement this rather than bind to a datadog only. And let users have an option in the configuration.

Reply:
Yes, let's try. Since there is not too much metrics for RR maybe will be not so painful. Ofc I will send you a feedback

Reply:
Only one thing, please, create a new FR ticket in the roadrunner-server/roadrunner repo.